design-proposals/sdn-ipam-secondary-nets.md

```diff
@@ -0,0 +1,490 @@
+
+#### Hot-unplug a VM interface
+Hot-unplugging an interface from a KubeVirt VM is a two step process, requiring
+a migration:
+1. the user mutates the VM.Spec.Template of the running VMI setting the
+   interface(s) as absent
+2. the user migrates the VM
+3. the interfaces marked as absent will not be templated on the destination pod
+   (i.e. the migration destination pod will not have those interfaces)
+4. KubeVirt dettaches the interface from the live VM
+
+To ensure the IPs allocated to hot-unplugged interfaces are returned to the
+pool, KubeVirt needs to delete the `IPAMClaim`, otherwise, they will only be
+garbage collected when the VM is deleted.
+
+Finally, OVN-Kubernetes must react to the deletion of the IPAMClaim, and return
+this IP address to its in-memory pool.
+
+### Migration considerations
+When migrating VMs to a new node there will be a few moments (seconds even)
+where both source and destination pods will be ready at the same time, having
+the same IP address configured. This will surely lead to some dropped packets.
+
+Providing seamless live-migration is not one of the goals of this feature - a
+follow up epic will focus on that, by binding the logical switch port for the
+pod to the node where the VM is available in "real time" - i.e. initially the
+port is bound to the src node, then it becomes bound on both nodes, then
+finally it is bound exclusively to the dst node (when the destination pod takes
+over). This feature is called OVN multi-chassis binding; Openstack uses it
+successfully, as indicated in
+[this talk](https://www.youtube.com/watch?v=ijZTMXAg-eI).
+
+### KubeVirt related changes
+
+#### Pod templating
+
+KubeVirt would need to template the pod accordingly - i.e. the
+`NetworkSelectionElements` must feature the `ipam-claim-reference` attribute to
+indicate to the IPAM CNI / CNI which `IPAMClaim` they should use to persist the
+IP allocation across reboots / migrations.
+
+Check the following network selection elements annotation from a KubeVirt
+launcher pod to see an example:
+
+```bash
+kubectl get pods <launcher pod name> -ojsonpath="{ @.metadata.annotations.k8s\.v1\.cni\.cncf\.io\/networks }" | jq
+[
+  {
+    "name": "tenantred",
+    "namespace": "default",
+    "interface": "pod16367aacb67",
+    "ipam-claim-reference": "vm-a.tenantred"
+  }
+]
+```
+
+#### Creating the IPAMClaim
+KubeVirt must create the IPAMClaims (and set finalizers / owner references) to
+properly ensure the VM's lifecycle is reflected in the allocation.
+
+There are two important sections to fill when creating the IPAMClaim reference:
+- the IPAMClaim name
+- the spec.network attribute
+
+The `IPAMClaim` name will be created using the following pattern:
+`<vm name>.<logical network name>`. The logical network name is the [name of the
+network](https://kubevirt.io/api-reference/main/definitions.html#_v1_network)
+in the KubeVirt API.
+
+The `IPAMClaim.spec.network` must be the name of network as per the
```

Comment:
it uses the network selection element. Seems the diagram is wrong, good catch.

Reply:
the CNI uses it when synchronizing its IP pool using `IPAMClaim`s. It needs to know to which network does the allocation belong.

Reply:
Do you mean in case the CNI crashes and restarts?

Reply:
I still don't get why the network name is needed on the IPAMClaim. If the IPAM CNI has an access to the pods and to the NADs it can get the network name.

Reply:
Added an explanation about this.